ethdk/src/types/brands.ts

```diff
@@ -0,0 +1,5 @@
+/** Generic Brand type to allow nominal typing */
+export type Brand<T, B> = T & { __brand: B }
```

Comment:
Decided to use type branding to discern between a private key and seed phrase. Similar concept 

sources/further reading
- chatgpt
- https://basarat.gitbook.io/typescript/main-1/nominaltyping - this link has some other solutions if anyone is interested
- https://github.com/kourge/ts-brand

Reply:
This is cool, I didn't know this was a thing.  I think this is a cool solution but I am torn on this because I think it may be overcomplicating the pk/seed phrase a little.  For example, I'm not a fan of having to cast the value that we are using in the config object like so.

```typescript
const account = await ExternallyOwnedAccount.createAccount({
      privateKeyOrSeedPhrase: privateKey as PrivateKey,
      network,
})
```

I did some thinking on this over the weekend because I was wondering how we could make this simpler.  I was thinking it would probably be best to have different interfaces for the EOA class.  Something like this

```typescript
type BaseConfig = {
  network?: Network
}

type PrivateKeyEoa = {
  privateKey: string
} & BaseConfig

type SeedPhraseEoa = {
  seedPhrase: string
} & BaseConfig

type EoaConfig = BaseConfig | PrivateKeyEoa | SeedPhraseEoa

static async createAccount(
    config: EoaConfig = {},
): Promise<ExternallyOwnedAccount> {
    if (isPrivateKey(config)) {
      console.log('Private Key:', config.privateKey)
    } else if (isSeedPhrase(config)) {
      console.log('Seed Phrase:', config.seedPhrase)
    }
}
```

I don't have a strong opinion, so if you want to stay with the nominal typing thats cool.  I was just trying to think of a way that we don't have to assert the type of a value we pass to the createAccount function or have a long key name.

Reply:
Hmm I found an issue with my above code.  It doesn't give a typescript error if you have both the seedPhrase and the privateKey like so
```typescript
{
      seedPhrase: 'test',
      privateKey: 'asdf',
      network,
}
```
I guess that's because typescript allows excess properties for objects

Reply:
Agree regarding added type brand adding complexity, casting and long variable names. As mentioned offline:

Changes:
- only allow one type at a time (no `privateKeyOrSeedPhrase`, or even a more abstract `key`). So we don't have to assert on the type and it's clear to the user what they are creating the account with, with either `createAccountFromSeedPhrase` or `createAccountFromPrivateKey`
- `createAccount` method at top level of ethdk is only concerned with sensible defaults. So it now only has a private key parameter, and no seed phrase.

This removes the branding, casting to `PrivateKey`, creates clear separation of concerns with the dedicated static methods on the EOA class, while also allowing a sensible default option through `createAccount`.

Reply:
> I guess that's because typescript allows excess properties for objects

The updated code still doesn't provide the behaviour we're describing here. e.g. I didn't get a compile error with the old parameter name of `privateKeyOrSeedPhrase`, and had to manually debug before I realised the issue in the test set up.

Would be good to find a nice solution to this. It's not hard to make a typo like `privateKeu` without noticing, so would be a good DX win

Reply:
We could add more type safety on the account config by creating a dedicated type/interface:

```ts
interface CreateAccountConfig {
  privateKey?: string
  network?: Network
}
```

This throws a compile error when typos are being passed in like `privateKeu`. Issue with this approach from testing it out, is it give less helpful type information and documentation when hovering over the methods:

before:
```ts
(method) ExternallyOwnedAccount.createAccount({ privateKey, network, }?: {
    privateKey?: string;
    network?: Network;
}): ExternallyOwnedAccount
```

after:
```ts
(method) ExternallyOwnedAccount.createAccount({ privateKey, network, }?: CreateAccountConfig): ExternallyOwnedAccount
```

Reply:
@JohnGuilding Is there a functional difference between the before and after?  I've always thought the before and after are the same.  The `after` may just be a little more readable since the config is broken out into an interface.  It could also be more reusable if that same CreateAccountConfig is used across multiple methods. But I think the type safety should be the same.

Reply:
Just tested again and yeah you're right, I made a mistake when testing this - I forgot to add a variable for the typo I made, which made me interpret the error wrong lol. So yeah think the only difference is readability